### PR TITLE
[BEAM-2858] Ensure that file deletion happens consistently, and only after table loads are completed

### PR DESCRIPTION
**Pull request by reuvenlax**
We do this by adding a GroupByKey between the BQ load and the file deletion. Once the @StableReplay proposal is implemented, we will replace with a solution that uses that instead.

R: @jkff

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java [hunk 1]
```diff
@@ -87,6 +88,28 @@ public Result(String filename, Long fileByteSize, DestinationT destination) {
       this.fileByteSize = fileByteSize;
       this.destination = destination;
     }
+
+    @Override
+    public String toString() {
+      return String.format("filename: %s fileByeSize: %d destination: %s",
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java]: There's a typo. Use MoreObjects.toStringHelper or ask IntelliJ to generate this for you, like in other classes?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 2]
```diff
@@ -489,6 +489,38 @@ public void processElement(ProcessContext c) throws Exception {
   }
 
   @Test
+  public void testWriteEmptyPCollection() throws Exception {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: Was there also a bug about writing empty collections that is addressed in this PR?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 3]
```diff
@@ -1142,12 +1180,20 @@ public void testWriteUnknown() throws Exception {
         .withDatasetService(datasetService);
     datasetService.createDataset("project-id", "dataset-id", "", "");
     Pipeline p = TestPipeline.create(bqOptions);
+
+    TableSchema schema = new TableSchema()
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: Why this change? Likewise for a couple of other changes below - they don't seem related to this PR.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 4]
```diff
@@ -2092,26 +2146,29 @@ public void testWriteTables() throws Exception {
...
-      }
-      tables.add(element.getValue());
-    }
-    assertEquals(expectedTempTables, tempTablesResult);
+    PAssert.thatMultimap(writeTablesOutput)
+        .satisfies(
+            new SerializableFunction<Map<TableDestination, Iterable<String>>, Void>() {
+              @Override
+              public Void apply(Map<TableDestination, Iterable<String>> input) {
+                assertEquals(input.keySet(), expectedTempTables.keySet());
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: This seems equivalent to assertEquals(HashMultimap.create(input), expectedTempTables).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java (hunk 1 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java]: > **jkff** wrote:
> There's a typo. Use MoreObjects.toStringHelper or ask IntelliJ to generate this for you, like in other classes?


Done.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java (hunk 3 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: > **jkff** wrote:
> Why this change? Likewise for a couple of other changes below - they don't seem related to this PR.


All of the TableSchema changes are indeed unneeded. I've removed them.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java (hunk 2 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: > **jkff** wrote:
> Was there also a bug about writing empty collections that is addressed in this PR?


I inherited this test from another PR. However we don't have any test coverage for writing an empty PCollection, so I left it in for the sake of coverage.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java (hunk 4 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: > **jkff** wrote:
> This seems equivalent to assertEquals(HashMultimap.create(input), expectedTempTables).


Except that doesn't compile, since input is not a Multimap. Also even if it did, it's unclear to me whether equals() would properly handle unordered lists of values (I think Multimap might just delegate to the underlying collection type).

Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: You're right about the equality - then you could do assertEquals(HashMultimap.create(input), HashMultimap.create(expectedTempTables)) - for HashMultimap, equality is defined in the way you need. Or you could just change expectedTempTables to explicitly be a HashMultimap.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java [hunk 5]
```diff
@@ -70,82 +85,133 @@
   private final WriteDisposition firstPaneWriteDisposition;
   private final CreateDisposition firstPaneCreateDisposition;
   private final DynamicDestinations<?, DestinationT> dynamicDestinations;
-  private Map<DestinationT, String> jsonSchemas = Maps.newHashMap();
+  private final List<PCollectionView<?>> sideInputs;
+  private final TupleTag<KV<TableDestination, String>> mainOutputTag;
+  private final TupleTag<String> temporaryFilesTag;
+
+
+  private class WriteTablesDoFn
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java]: Can you simplify the diff by preserving the order of functions? - i.e. put WriteTablesDoFn between constructor and expand.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java [hunk 6]
```diff
@@ -70,82 +85,133 @@
...
+    // Garbage collect temporary files.
+    // We mustn't start garbage collecting files until we are assured that the WriteTablesDoFn has
+    // succeeded in loading those files and won't be retried. Otherwise, we might fail part of the
+    // way through deleting temporary files, and retry WriteTablesDoFn. This will then fail due
+    // to missing files, causing either the entire workflow to fail or get stuck (depending on how
+    // the runner handles persistent failures).
+    writeTablesOutputs
+        .get(temporaryFilesTag)
+        .setCoder(StringUtf8Coder.of())
+        .apply(WithKeys.<Void, String>of((Void) null))
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java]: > **jkff** wrote:
> Explain why we group with a void key (this is because GCTemporaryFiles takes an Iterable, right? Is there any reason why we want it to continue taking an Iterable, as opposed to taking single elements and batching the deletions within a bundle?)
> 
> And also explain why we window into global windows and trigger "as fast as possible" - is this because we want file deletion to happen as fast as possible? But isn't the input already globally windowed? Perhaps you need Window.configure().triggering... to emphasize that you're not changing the windowing strategy?


It's because we need a GBK, so we need _some_ key to give.

Also Window.configure().triggering - is that allowed? It's not really documented anywhere.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java (hunk 5 above)
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java]: > **jkff** wrote:
> Can you simplify the diff by preserving the order of functions? - i.e. put WriteTablesDoFn between constructor and expand.


Done.
